schema/gen/go/genStructReprTuple.go

```diff
@@ -0,0 +1,443 @@
+
+	// Count how many trailing fields are optional.
+	//  The 'Done' predicate gets more complex when in the trailing optionals.
+	fields := g.Type.Fields()
+	fieldCount := len(fields)
+	beginTrailingOptionalField := fieldCount
+	for i := fieldCount - 1; i >= 0; i-- {
+		if !fields[i].IsOptional() {
+			break
+		}
+		beginTrailingOptionalField = i
+	}
+	haveTrailingOptionals := beginTrailingOptionalField < fieldCount
+
+	// Now: finally we can get on with the actual templating.
+	doTemplate(`
+		func (n *_{{ .Type | TypeSymbol }}__Repr) ListIterator() ipld.ListIterator {
+			{{- if .HaveTrailingOptionals }}
+			end := {{ len .Type.Fields }}`+
+		func() string { // this next part was too silly in templates due to lack of reverse ranging.
+			v := "\n"
+			for i := fieldCount - 1; i >= beginTrailingOptionalField; i-- {
+				v += "\t\t\tif n." + g.AdjCfg.FieldSymbolLower(fields[i]) + ".m == schema.Maybe_Absent {\n"
+				v += "\t\t\t\tend = " + strconv.Itoa(i) + "\n"
+				v += "\t\t\t} else {\n"
+				v += "\t\t\t\tgoto done\n"
+				v += "\t\t\t}\n"
+			}
+			return v
+		}()+`done:
+			return &_{{ .Type | TypeSymbol }}__ReprListItr{n, 0, end}
+			{{- else}}
+			return &_{{ .Type | TypeSymbol }}__ReprListItr{n, 0}
+			{{- end}}
+		}
+
+		type _{{ .Type | TypeSymbol }}__ReprListItr struct {
+			n   *_{{ .Type | TypeSymbol }}__Repr
+			idx int
+			{{if .HaveTrailingOptionals }}end int{{end}}
+		}
+
+		func (itr *_{{ .Type | TypeSymbol }}__ReprListItr) Next() (idx int, v ipld.Node, err error) {
+			if itr.idx >= {{ len .Type.Fields }} {
+				return -1, nil, ipld.ErrIteratorOverread{}
+			}
+			switch itr.idx {
+			{{- range $i, $field := .Type.Fields }}
+			case {{ $i }}:
+				idx = itr.idx
+				{{- if $field.IsOptional }}
+				if itr.n.{{ $field | FieldSymbolLower }}.m == schema.Maybe_Absent {
+					return -1, nil, ipld.ErrIteratorOverread{}
+				}
+				{{- end}}
+				{{- if $field.IsNullable }}
+				if itr.n.{{ $field | FieldSymbolLower }}.m == schema.Maybe_Null {
+					v = ipld.Null
+					break
+				}
+				{{- end}}
+				{{- if $field.IsMaybe }}
+				v = itr.n.{{ $field | FieldSymbolLower}}.v.Representation()
+				{{- else}}
+				v = itr.n.{{ $field | FieldSymbolLower}}.Representation()
+				{{- end}}
+			{{- end}}
+			default:
+				panic("unreachable")
+			}
+			itr.idx++
+			return
+		}
+		{{- if .HaveTrailingOptionals }}
+		func (itr *_{{ .Type | TypeSymbol }}__ReprListItr) Done() bool {
+			return itr.idx >= itr.end
+		}
+		{{- else}}
+		func (itr *_{{ .Type | TypeSymbol }}__ReprListItr) Done() bool {
+			return itr.idx >= {{ len .Type.Fields }}
+		}
+		{{- end}}
+
+	`, w, g.AdjCfg, struct {
+		Type                  *schema.TypeStruct
+		HaveTrailingOptionals bool
+	}{
+		g.Type,
+		haveTrailingOptionals,
+	})
+}
+
+func (g structReprTupleReprGenerator) EmitNodeMethodLength(w io.Writer) {
+	// This is fun: it has to count down for any unset optional fields.
```

Comment:
don't we have special rules for optional fields in tuple representation? like, only the last one can be optional for it to make much sense?

Reply:
Yep.  There's some comments about that at the top of the file.  (There can be several optionals... as long as they're all consecutive, and the matching is greedy.)

I think I was lazy in this snippet: some of the code is very similar to code used in optional-value-presense-counting in structs representations as maps, and I just didn't bother to rewrite it to use the assumptions we're allowed here.

(Optimistically, I was thinking we might be able to make them share code entirely, but I experienced "mental OOM" trying to get there, so this is the non-DRY'd shrapnel.  There will be a *lot* of stuff like this that deserves revisit in the near future: things will either turn out to be very "obvious" candidates for DRYing, or, very much deserving of further specialization -- but I'm frequently giving up trying to tell the difference as I chase feature-completeness right now.  Mostly, I'm trying to at least mark them in comments, but I'm sure I often fail.)